components/supervisor-api/notification.proto

```diff
@@ -11,60 +11,93 @@ import "google/api/annotations.proto";
-    }
-
-    // Subscribe to notifications. Typically called by the IDE.
-    rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
-        option (google.api.http) = {
-            get: "/v1/notification/subscribe"
-        };
-    }
-
-    // Report a user's choice as a response to a notification. Typically called by the IDE.
-    rpc Respond(RespondRequest) returns (RespondResponse) {
-        option (google.api.http) = {
-            post: "/v1/notification/respond"
-        };
-    }
+  // Prompts the user and asks for a decision. Typically called by some external
+  // process. If the list of actions is empty this service returns immediately,
+  // otherwise it blocks until the user has made their choice.
+  rpc Notify(NotifyRequest) returns (NotifyResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/notify"
+    };
+  }
+
+  // Subscribe to notifications. Typically called by the IDE.
+  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
+    option (google.api.http) = {
+      get : "/v1/notification/subscribe"
+    };
+  }
+
+  // Report a user's choice as a response to a notification. Typically called by
+  // the IDE.
+  rpc Respond(RespondRequest) returns (RespondResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/respond"
+    };
+  }
+
+  // Called by the notifications clients to inform supervisor about
+  // which is the latest client actively used by the user.
+  // Useful for gp open / gp preview to trigger actions in the right client
+  rpc SetActiveClient(SetActiveClientRequest)
+      returns (SetActiveClientResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/active-client"
+    };
+  }
+
+  rpc GetActiveClient(GetActiveClientRequest)
+      returns (GetActiveClientResponse) {
+    option (google.api.http) = {
+      get : "/v1/notification/active-client"
+    };
+  }
 }
 
 message NotifyRequest {
-    enum Level {
-        ERROR = 0;
-        WARNING = 1;
-        INFO = 2;
-    }
-    Level level = 1;
-    string message = 2;
-    // if actions are empty, Notify will return immediately
-    repeated string actions = 3;
+  enum Level {
+    ERROR = 0;
+    WARNING = 1;
+    INFO = 2;
+  }
+  Level level = 1;
+  string message = 2;
+  // if actions are empty, Notify will return immediately
+  repeated string actions = 3;
 }
 
 message NotifyResponse {
-    // action chosen by the user or empty string if cancelled
-    string action = 1;
+  // action chosen by the user or empty string if cancelled
+  string action = 1;
 }
 
 message SubscribeRequest {}
 
 message SubscribeResponse {
-    uint64 requestId = 1;
-    NotifyRequest request = 2;
+  uint64 requestId = 1;
+  NotifyRequest request = 2;
 }
 
 message RespondRequest {
-    uint64 requestId = 1;
-    NotifyResponse response = 2;
+  uint64 requestId = 1;
+  NotifyResponse response = 2;
+}
+
+message RespondResponse {}
+
+enum ActiveClient {
+  code = 0;
```

Comment:
I think identity should be just a string, that it can be extended by 3rd party IDEs later.

Most likely it also won't be `code`, but more like someting like `code-${uuid}`, where is an uuid is session ID of VS Code window.

Reply:
We should also allow clients to unset active, i.e. if we detect that VS Code window is not focused anymore then it should indicate it to supervisor.

Reply:
@akosyakov How about https://github.com/gitpod-io/gitpod/issues/8191#issuecomment-1096110343 this?